dataframe/at_indexer.go

```diff
@@ -0,0 +1,97 @@
+package dataframe
+
+import (
+	"fmt"
+
+	"go.starlark.net/starlark"
+)
+
+// AtIndexer is returned by dataframe.At and can index to any cell
+type AtIndexer struct {
+	owner *DataFrame
+}
+
+// compile-time interface assertions
+var (
+	_ starlark.Value     = (*AtIndexer)(nil)
+	_ starlark.Mapping   = (*AtIndexer)(nil)
+	_ starlark.HasSetKey = (*AtIndexer)(nil)
+)
+
+// NewAtIndexer returns a new AtIndexer with the text values and name
+func NewAtIndexer(owner *DataFrame) *AtIndexer {
+	return &AtIndexer{owner: owner}
+}
+
+// Freeze prevents the index from being mutated
+func (ai *AtIndexer) Freeze() {
+}
+
+// Hash cannot be used with AtIndexer
+func (ai *AtIndexer) Hash() (uint32, error) {
+	return 0, fmt.Errorf("unhashable: %s", ai.Type())
+}
+
+// String returns the index as a string
+func (ai *AtIndexer) String() string {
+	return fmt.Sprintf("AtIndexer(...)")
+}
+
+// Truth converts the index into a bool
+func (ai *AtIndexer) Truth() starlark.Bool {
+	return true
+}
+
+// Type returns the type as a string
+func (ai *AtIndexer) Type() string {
+	return fmt.Sprintf("%s.AtIndexer", Name)
+}
+
+// Get returns the value at the given key, which must be an int pair
+func (ai *AtIndexer) Get(key starlark.Value) (starlark.Value, bool, error) {
```

Comment:
This is a good example of the "three audience" discussion we had on a call. Where those audiences are:

1. Starlark script authors (userland)
2. The computer
3. Go developers.

This function gives me questions for the third audience. This is an exported method on an exported type, which means it's part of the API go developers will see, but it accepts starlark values. To me we can steer the Go Developer audience in one of two directions:

1. We expect go developers need this method, and will need to construct `starlark.Value` arguments to call it.
2. We _don't_ expect Go Developers to call this method, or construct `starlark.Value`'s.

In the first scenario we might need to consider exporting helper functions for getting from go types to Starlark types. In the second we should unexport this method and any other method that accepts arguments from the `starlark` package.

I'm personally a fan of the latter, and would be in favor of taking a "progressive enhancement" approach. Deal with existing API's later and just focus on not introducing exported methods that accept `starlark` package arguments where possible. If you're into that approach we should un-export this method, and reserve an exported `Get` method if Go developers need it. It's signature would look something like:

```go
func (ai *AtIndexer) Get(i,j int) (interface{}, bool, error)
```

But I don't think we actually implement that method until the Go developer audience can articulate a case where it's needed. Might be able to audit our existing test suite for clues

Reply:
Can't change the signature of this function, it is used by the `Mapping` interface for when a user does `cell = df.at[3,4]`.

I agree that the general question of what methods are designed for golang and which are designed for starlark's engine needs some sort of convention, likely just a naming scheme. I'm still thinking of how best to do this, and consider it a work in progress. Unfortunately just implementing the interfaces that we need, in order to serve starlark, must take priority.

Reply:
sounds good, and that's definitely the right priority. We should make things work for starlark users at the expense of complexity for go users whenever that tradeoff is necessary